src/main/java/io/supertokens/storage/mongodb/queries/JWTSigningQueries.java

```diff
@@ -0,0 +1,135 @@
+        List<JWTSigningKeyInfo> result = new ArrayList<>();
+
+        /*
+         * Schema for the collection is
+         * {
+         *  _id: string, (key id)
+         *  key_string: string,
+         *  created_at: long,
+         *  algorithm: string,
+         * }
+         *
+         * created_at should only be used to determine the key that was added to the database last, it should not be
+         * used to determine the validity or lifetime of a key. While the assumption that created_at refers to the time
+         * the key was generated holds true for keys generated by the core, it is not guaranteed when we allow user defined
+         * keys in the future.
+         */
+        try (MongoCursor cursor = collection.find().sort(Sorts.descending("created_at")).iterator()) {
+            while (cursor.hasNext()) {
+                Document currentDoc = (Document) cursor.next();
+                result.add(JWTSigningKeyInfoRowMapper.getInstance().mapOrThrow(currentDoc));
+            }
+        }
+
+        return result;
+    }
+
+    private static class JWTSigningKeyInfoRowMapper implements RowMapper<JWTSigningKeyInfo, Document> {
+        private static final JWTSigningKeyInfoRowMapper INSTANCE = new JWTSigningKeyInfoRowMapper();
+
+        private JWTSigningKeyInfoRowMapper() {
+        }
+
+        private static JWTSigningKeyInfoRowMapper getInstance() {
+            return INSTANCE;
+        }
+
+        @Override
+        public JWTSigningKeyInfo map(Document result) throws Exception {
+            String keyId = result.getString("_id");
+            String keyString = result.getString("key_string");
+            long createdAt = result.getLong("created_at");
+            String algorithm = result.getString("algorithm");
+
+            if (keyString.contains("|")) {
+                return new JWTAsymmetricSigningKeyInfo(keyId, createdAt, algorithm, keyString);
+            } else {
+                return new JWTSymmetricSigningKeyInfo(keyId, createdAt, algorithm, keyString);
+            }
+        }
+    }
+
+    public static boolean setJWTSigningKeyInfoIfNoKeyForAlgorithmExists_Transaction(Start start, JWTSigningKeyInfo keyInfo) throws StorageQueryException {
+        MongoDatabase client = ConnectionPool.getClientConnectedToDatabase(start);
+        MongoCollection collection = client.getCollection(Config.getConfig(start).getJWTSigningKeysCollection());
+
+        Document toInsertIfNoneFound = new Document(
+                "$setOnInsert",
+                new Document("_id", keyInfo.keyId)
+                    .append("key_string", keyInfo.keyString)
+                    .append("created_at", keyInfo.createdAtTime)
+                    .append("algorithm", keyInfo.algorithm)
+                );
+
+        FindOneAndUpdateOptions options = new FindOneAndUpdateOptions();
+        options.returnDocument(ReturnDocument.AFTER);
```

Comment:
What does this option do?

Reply:
By default the query returns the original document if this is set to `ReturnDocument.BEFORE`, when set to `ReturnDocument.AFTER` it will return the new document if one is updated (which in our case means one is inserted since we do `setOnInsert`. 

In our case this means that if the database had no keys for that algorithm, the query would return`null` if that flag is set to `ReturnDocument.BEFORE`. In this case we cannot ensure that the write was successful. When set to `ReturnDocument.AFTER`, if the database had no keys for that algorithm the query would return the newly inserted document